README.md
# picker [![spm version](http://moekit.com/badge/picker)](http://spmjs.io/package/picker)

---

## Usage
```js
var Picker = require("picker");
var pickerDate = new Picker(options);

// use Picker
```

只支持在wap端使用

## API

### API设计参考
picker的api参考淘宝f7的api  
http://framework7.taobao.org/docs/picker.html#picker-methods-properties  
而picker中列实例api则基本沿用picker-core的  
https://github.com/MoeKit/picker-core

### 实例化选项options
+ input: `String` 用于触发选择器和值显示，input的id，需带#号
+ container: `String` 嵌入页面盒子，盒子id，需带#号 *当此值存在时，对应选择器弹出框各项功能将被取消，如onOpen，onClose*
+ itemsNumber: `Number` 选择器列的可视行数，单位px, 默认7
+ itemHeight: `Number` 选择器列的行高，单位px, 默认30
+ formatValue: `Function` 所有列的值的格式化函数，此函数可重新定义值的格式，值的格式默认'val1 val2 val3 ...', ，格式化函数参数有两个：*[optional]*，函数需return一个值回来
	+ picker: 当前实例（后面详细说明）
	+ values: `Array` 所有列的值的集合
+ onChange: 列选项变更的回调行数，回调参数有三个: *[optional]*
	+ picker: 当前实例（后面详细说明）
	+ current 当前选中的值
	+ prev 之前的值
+ onOpen: (内部dialog实例化参数)选择器被打开的回调函数，回调参数有一个: *[optional]*
	+ picker: 当前实例（后面详细说明）
+ onClose: (内部dialog实例化参数)选择器被关闭的回调函数，回调参数有一个: *[optional]*
	+ picker: 当前实例（后面详细说明）
+ toolbarTemplate: 重新定制选择器弹出框顶部的toolbar的html，默认只有右上角有一个done按钮（用于关闭）, 默认html：
```html
<div class="picker-toolbar">
	<a href="javascript:void(0);" class="picker-toolbar-right">Done</a>
</div>
```
如需保留右上角事件，可copy上面的html进行增加修改~, 至于新的toolbar事件可在onOpen回调函数中绑定
+ cols：`[Object]` 一个对象数组，每一个对象代表一个选项列，对象的key有：
	+ width: `Number` 选项列的宽度，单位px
	+ textAlign: `String` 选项列的水平对齐方式，三种：left,right,center
	+ values: `Array` 选项列的实际值
	+ displayValues: `Array` 选项列的显示值，不存在该key值时，实际值充当显示值
	+ onChange: 细节同上（实例化选项的onChange）, 此处优先级大于实例化选项的onChange
	+ divider: `Boolean` 是否启用选项列的分隔符，true为启用
	+ content: `String` 分隔符的内容，启用分隔符才有效

## 实例方法与属性

### 属性
+ params `Object` 实例初始化的参数
+ cols `[Instance]` 每一列的实例(详解请看下面)
+ container `DOM` 存放当前实例的HTML容器
+ value `Array` 选中的值
+ dialog `Instance` 弹出框实例
	+ updateInputPosition `function` 更新input的位置，（input初始化是隐藏时导致位置异常，当设置为显示时需要调用此事件更新input位置）
	+ show `function` 等同于 picker.open
	+ hide `function` 等同于 picker.close

### 方法
+ open 显示选择器弹出框
+ close 隐藏选择器弹出框
+ setValue(values) 设置所有列的值
+ destroy 清空实例和事件

## cols列实例方法与属性

### 属性
+ values `Array` 实际列值
+ currentValue 当前值
+ prevValue 之前的值

### 方法
+ getValue() 获取当前的选中值
+ replaceValues `Function` 重新渲染选择器的选项，函数带两个参数：
	+ values `Array` 实际值
	+ displayValues `Array` 显示值（可选）
+ select(value) 通过值选中一项
	+ @param value {string|number}
+ selectByIndex(index) 通过索引选择一项
	+ @param index {number} 非负索引值

## dialog实例与方法

### 属性
+ params `Object` 实例初始化的参数
+ container `DOM` 存放当前实例的HTML容器

### 方法
+ updateInputPosition `Function` 更新input相对位置
+ show `Function` 显示弹出框
+ hide `Function` 隐藏弹出框

## 单元测试

HISTORY.md
# History

---

## 0.0.1

`new` It is the first version of picker.

## 0.0.2

`new` 实例结构，补全文档

## 0.0.3

`improved` input元素事件从picker迁入picker-dialog, 以保证picker能更好的处理业务逻辑;

`improved` simulate-click从HTMLElement的prototype移除，重新定制并导出到window.__tap__和window.__untap__;  

`fixed` 修正单元测试

## 0.0.4

`new`   打开picker时自动读取内容到input上
`new`   新增deom-日期（年月日）
`fixed` 修正input相对body高度
`fixed` 修正basescroller-scroller的宽度，使之宽度计算正常

tests/index-spec.js
var expect = require('expect.js');
var picker = require('../index');

describe('picker', function() {

  it('normal usage', function() {

  });

});


tests/picker-spec.js
var expect = require('expect.js');
var Picker = require('../index');
var input = document.createElement('input');
input.id = 'JS_picker';
input.type = 'text';
input.style.display = 'none';
document.body.appendChild(input);
var input = document.createElement('input');
input.id = 'JS_picker_destroy';
input.type = 'text';
input.style.display = 'none';
document.body.appendChild(input);
var div = document.createElement('div');
div.id = 'JS_container';
div.style.display = 'none';
document.body.appendChild(div);

describe('picker', function() {

    describe('参数', function() {
        it('是json格式, key有input：对应input标签的id（含#）', function() {
            var picker = new Picker({
                'input': '#JS_picker'
            });
        });

        it('itemNumber：可视选项的个数', function() {
            var picker = new Picker({
                'input': '#JS_picker',
                'itemNumber': 7
            });
        });

        it('itemHeight: 选项的高度', function() {
            var picker = new Picker({
                'input': '#JS_picker',
                'itemHeight': 30
            });
        });

        it('cols: 每一列选项的值', function() {
            var picker = new Picker({
                'input': '#JS_picker',
                'cols': [{
                    'values': [1, 2, 3, 4, 5, 6]
                }]
            });
        });

        it('toolbarTemplate: 顶部toolbar模板', function() {
            var picker = new Picker({
                'input': '#JS_picker',
                'toolbarTemplate': '<div>test</div>'
            });
        });

        it('formatValue: 选项改变后值输出的格式', function() {
            var picker = new Picker({
                'input': '#JS_picker',
                'formatValue': function(picker, values) {
                    return values.join(" ");
                }
            });
        });

        it('onChange: 选项改变时执行的事件', function() {
            var picker = new Picker({
                'input': '#JS_picker',
                'onChange': function(picker, current, old) {
                    console.log(current, old);
                }
            });
        });

        it('onOpen: 弹出框显示时执行的事件', function() {
            var picker = new Picker({
                'input': '#JS_picker',
                'onOpen': function(picker) {

                }
            });
        });

        it('onClose: 弹出框隐藏时执行的事件', function() {
            var picker = new Picker({
                'input': '#JS_picker',
                'onClose': function(picker) {

                }
            });
        });

        it('container: picker选项存放的地方，值为id（带#）', function() {
            var picker = new Picker({
                'input': '#JS_picker',
                'container': '#JS_container'
            });
        });

    });

	describe('事件', function() {

		it('close：隐藏弹出框事件', function(){
			var picker = new Picker({
	            'input': '#JS_picker'
	        });
            picker.close();
		});

		it('open：显示弹出框事件', function(){
			var picker = new Picker({
	            'input': '#JS_picker'
	        });
            picker.open();
            picker.close();
		});

		it('setValue：设置所有列值', function(){
			var picker = new Picker({
	            'input': '#JS_picker'
	        });
            picker.setValue([]);
		});

		it('destroy：设置所有列值', function(){
			var picker = new Picker({
	            'input': '#JS_picker_destroy'
	        });
            picker.destroy();
		});

	});

	describe('实例属性cols列的事件', function() {
		
		it('getValue：获取当前列的选中的值', function(){
			var picker = new Picker({
	            'input': '#JS_picker',
	            'cols': [
			        {
			            "values": [1,2,3,4,"4s",5,"5c","5s",6,"6s",7]
			        }
			    ]
	        });
            picker.cols[0].getValue();
		});

		it('replaceValues：重新渲染选择器选项', function(){
			var picker = new Picker({
	            'input': '#JS_picker',
	            'cols': [
			        {
			            "values": [1,2,3,4,"4s",5,"5c","5s",6,"6s",7]
			        }
			    ]
	        });
            picker.cols[0].replaceValues([23,2,345,456,12,312,4]);
		});

		it('select：通过值选中一项', function(){
			var picker = new Picker({
	            'input': '#JS_picker',
	            'cols': [
			        {
			            "values": [1,2,3,4,"4s",5,"5c","5s",6,"6s",7]
			        }
			    ]
	        });
            picker.cols[0].select(3);
		});

		it('selectByIndex：通过索引选择一项', function(){
			var picker = new Picker({
	            'input': '#JS_picker',
	            'cols': [
			        {
			            "values": [1,2,3,4,"4s",5,"5c","5s",6,"6s",7]
			        }
			    ]
	        });
            picker.cols[0].select(3);
		});

	});

    describe('容错处理', function() {
        it('参数为空', function() {
            var pickerq = new Picker();
        });

        it('参数为空的json', function() {
            var pickerw = new Picker({});
        });

        it('参数为json，且input对应的值不存在', function() {
            var picker = new Picker({
                'input': '#JS_test'
            });
        });

        it('参数为json，且container对应的值不存在', function() {
            var picker = new Picker({
                'input': '#JS_picker',
                'container': '#JS_test'
            });
        });
    });
});


tests/picker-dialog-spec.js
var expect = require('expect.js');
var PickerDialog = require('../src/picker-dialog.js');
var input = document.createElement('input');
input.id = 'JS_PickerDialog';
input.type = 'text';
input.style.display = 'none';
document.body.appendChild(input);

describe('picker-dialog', function() {

    describe('参数', function() {
        it('是json格式, key有input：对应input标签的id（含#）', function() {
            var picker = new PickerDialog({
                'input': '#JS_PickerDialog'
            });
        });

        it('innerHTML, dialog的内容', function() {
            var picker = new PickerDialog({
                'input': '#JS_PickerDialog',
                'innerHTML': "test/test/test"
            });
        });

        it('onOpen: 弹出框显示时执行的事件', function() {
            var picker = new PickerDialog({
                'input': '#JS_PickerDialog',
                'onOpen': function() {

                }
            });
        });

        it('onClose: 弹出框隐藏时执行的事件', function() {
            var picker = new PickerDialog({
                'input': '#JS_PickerDialog',
                'onClose': function() {

                }
            });
        });
    });

    describe('事件', function() {
        it('show事件, 显示弹出框', function() {
            var pickerDialog = new PickerDialog({
                'input': '#JS_PickerDialog'
            });
            pickerDialog.show();
        });

        it('hide事件, 显示弹出框', function() {
            var pickerDialog = new PickerDialog({
                'input': '#JS_PickerDialog'
            });
            pickerDialog.hide();
        });

        it('html事件, 改变dialog的内容', function() {
            var pickerDialog = new PickerDialog({
                'input': '#JS_PickerDialog'
            });
            pickerDialog.html("真的能改变吗");
        });

        it('updateInputPosition事件, 更新input的位置', function() {
            var pickerDialog = new PickerDialog({
                'input': '#JS_PickerDialog'
            });
            pickerDialog.updateInputPosition();
        });
        
    });

    describe('容错处理', function() {
        it('参数是随意对象、数字、字符、数组等等', function() {
            var pickerDialog = new PickerDialog();
            var pickerDialog = new PickerDialog({
                'key': 'value'
            });
            var pickerDialog = new PickerDialog(['i am array']);
            var pickerDialog = new PickerDialog('i am string');
            var pickerDialog = new PickerDialog(110112114119);
        });
    });
    
});


src/picker-dialog.js
'use strict';
require('./picker-dialog.css');
var Eventor = require('eventor');
var F_tap = require('./simulate-click.js');
var PickerDialog = function(option){
	this.params = {};
	if( Object.prototype.toString.call(option) === '[object Object]' ){
		this.params = {
            "input": option.input || "",
            "container": document.querySelector(option.input) || "",
            "innerHTML": option.innerHTML || "",
            "onOpen": option.onOpen || function(){},
            "onClose": option.onClose || function(){},
            "_open": option._open || function(){},
            "_close": option._close || function(){}
        };
        this.input = document.querySelector(option.input);
	}
	
	// input不存在提示处理
	if (!this.params || !this.params.input) {
        console.error('picker-dialog: input对应的dom对象不存在');
        return false;
    }
    if( !this.params.container || !this.params.container ){
        console.error('picker-dialog: container对应的dom对象不存在');
        return false;
    }
	// 透明黑色遮罩
	if( !!document.querySelectorAll('.picker-mask').length<=0 ){
		var divMask = document.createElement('a');
		divMask.className = 'picker-mask';
		divMask.href = 'javascript:void(0);';
		document.body.appendChild(divMask);
	}
	var div = document.createElement('div');
	div.className = 'picker-dialog';
	document.body.appendChild(div);
	this.mask = document.querySelector(".picker-mask");
	this.container = document.querySelectorAll(".picker-dialog");
	this.container = this.container[ this.container.length-1 ];
	this.container.innerHTML = this.params.innerHTML;
	this._hackInputFocus();
    this._bindEvents();
	option = null;
	return this;
}
Eventor.mixTo(PickerDialog);
PickerDialog.prototype._hackInputFocus = function() {
    // 兼容安卓微信，在input上面增加一层遮罩层，安卓微信没法禁用不可编辑
    var _inputMask = document.createElement('div');
    var _input = this.params.container;
    var _parent = _input.parentNode;
    var _inputMaskPage = document.getElementById(_input.getAttribute('id') + 'Mask');
    if( _inputMaskPage ){
        _inputMaskPage.style.width = (_input.offsetWidth + _input.clientLeft * 2) + 'px';
        _inputMaskPage.style.height = (_input.offsetHeight + _input.clientTop * 2) + 'px';
        _inputMaskPage.style.position = 'absolute';
        _inputMaskPage.style.left = _input.offsetLeft + "px";
        _inputMaskPage.style.top = _input.offsetTop + "px";
    } else {
        _inputMask.style.width = (_input.offsetWidth + _input.clientLeft * 2) + 'px';
        _inputMask.style.height = (_input.offsetHeight + _input.clientTop * 2) + 'px';
        _inputMask.style.position = 'absolute';
        _inputMask.style.left = _input.offsetLeft + "px";
        _inputMask.style.top = _input.offsetTop + "px";
        _inputMask.setAttribute('id', _input.getAttribute('id') + 'Mask');
        _parent.insertBefore(_inputMask, _input);
        this.input = document.getElementById(_input.getAttribute('id') + 'Mask');
    }
    this._bindEventInput();
    return this;
};
PickerDialog.prototype.updateInputPosition = function(){
    this._hackInputFocus();
}
PickerDialog.prototype._bindEventInput = function() {
    // 设置input被触发的事件
    var _this = this;
    F_tap.untap( this.input );
    var inputEvent = function(){
        _this.show();
        // 如果input被picker遮挡到，则滚动input至可视区域
        (function() {
            var clientHeight = document.documentElement.clientHeight; //  浏览器高度
            var scrollTop = document.body.scrollTop; // 滚动高度
            var pickerHeight = _this.container.offsetHeight; // picker高度
            var inputTop = _this.input.getBoundingClientRect().top + scrollTop; // input相对body高度
            var inputHeight = _this.input.offsetHeight; // input高度
            if (inputTop - scrollTop + inputHeight > clientHeight - pickerHeight || inputTop - scrollTop + inputHeight < inputHeight) {
                _this.scrollAnimate(inputTop - (clientHeight - pickerHeight) / 2, 400);
            } else {
            }
        }());
    };
    F_tap.tap( this.input, inputEvent);
    F_tap.tap( this.params.container, inputEvent);
    return this;

};
// input输入框如果不在可视区域内，将其滑动到可视区域
PickerDialog.prototype.scrollAnimate = function(position, timestamp) {
    var _this = this;
    var needs = (document.body.scrollTop - position) / timestamp * 25;
    var old = -1;
    _this.interval = setInterval(function() {
        if (Math.abs(document.body.scrollTop - position) < Math.abs(needs)) {
            clearInterval(_this.interval);
        }
        old = document.body.scrollTop;
        document.body.scrollTop = old - needs;
        if( old == document.body.scrollTop ){
        	clearInterval(_this.interval);
        }
    }, 8);
    return this;
};
PickerDialog.prototype._bindEvents = function(){
	var _this = this;
	function triggerClick(e){
		_this.hide();
		_this.emit('close');
	}
	F_tap.tap( this.mask, triggerClick );
	this.container.addEventListener("touchmove", function(e){
		e.stopPropagation();
		e.preventDefault();
		return false;
	}, false);
	// this.mask.tap(triggerClick);
	//this.setClick(this.mask, triggerClick);
	return this;
}
PickerDialog.prototype.show = function(){
	var _this = this;
	_this.mask.classList.add("show");
	_this.container.classList.add("modal-in");
    _this.params._open && _this.params._open(this);
	_this.params.onOpen && _this.params.onOpen(this);
	return this;
}
PickerDialog.prototype.hide = function(){
	var _this = this;
	_this.mask.classList.remove("show");
	_this.container.classList.remove("modal-in");
    _this.params._close && _this.params._close(this);
	_this.params.onClose && _this.params.onClose(this);
	return this;
}
PickerDialog.prototype.html = function(html){
	var _this = this;
	this.container.innerHTML = html;
	return this;
}
module.exports =  PickerDialog;

src/simulate-click.js
'use strict';
var Event = {
    tap: function(element, callback){
        if( !element ) return console.error("tap对象不能为空");
        element.__tap = {};
        element.__tap.event = {
            start: function(e) {
                e.stopPropagation();
                element.__tap.clickabled = true;
                element.__tap.starttime = e.timeStamp;
                element.__tap.startPageX = e.changedTouches[0].pageX;
                element.__tap.startPageY = e.changedTouches[0].pageY;
            },
            move: function(e) {
                if (Math.abs(e.changedTouches[0].pageX - element.__tap.startPageX) >= 5 || 
                    Math.abs(e.changedTouches[0].pageY - element.__tap.startPageY) >= 5) {
                    element.__tap.clickabled = false;
                }
            },
            end: function(e) {
                e.stopPropagation();
                e.preventDefault();
                if (e.timeStamp - element.__tap.starttime > 30 && 
                    e.timeStamp - element.__tap.starttime < 300 && 
                    element.__tap.clickabled ) {
                    // setTimeout(function() {
                        if (!!callback) {
                            callback(e);
                        }
                    // }, 0);
                }
            },
            click: function(e) {
                e.stopPropagation();
                callback && callback(e);
            }
        }
        if (!!navigator.userAgent.match(/AppleWebKit.*Mobile.*/)) {
            element.addEventListener('touchstart', element.__tap.event.start, false);
            element.addEventListener('touchmove', element.__tap.event.move, false);
            element.addEventListener('touchend', element.__tap.event.end, false);
        } else {
            element.addEventListener('click', element.__tap.event.click, false);
        }
        return element;
    },
    untap: function(element){
        if( !element ) return console.error("untap对象不能为空");
        element.__tap = element.__tap || {};
        if (!!navigator.userAgent.match(/AppleWebKit.*Mobile.*/) && !!element.__tap.event ) {
            element.__tap.event.start && element.removeEventListener('touchstart', element.__tap.event.start, false);
            element.__tap.event.move && element.removeEventListener('touchmove', element.__tap.event.move, false);
            element.__tap.event.end && element.removeEventListener('touchend', element.__tap.event.end, false);
        } else if(!!element.__tap.event) {
            element.__tap.event.click && element.removeEventListener('click', element.__tap.event.click, false);
        }
        return element;
    }

};
window.__tap__ = Event.tap;
window.__untap__ = Event.untap;
module.exports = Event;

// module.exports = function() {
//     HTMLElement.prototype.tap = function(fn) {
//         var _this = this;
//         this.__t = {};
//         this.tevent = {
//             tstart: function(e) {
//                 this.clickabled = true;
//                 e.stopPropagation();
//                 _this.__t.start = e.timeStamp;
//                 _this.__t.startEvent = e;
//             },
//             tmove: function(e) {
//                 if (Math.abs(e.changedTouches[0].pageX - _this.__t.startEvent.changedTouches[0].pageX) > 10 || Math.abs(e.changedTouches[0].pageY - _this.__t.startEvent.changedTouches[0].pageY) > 10) {
//                     this.clickabled = false;
//                 }
//             },
//             tend: function(e) {
//                 e.stopPropagation();
//                 if (e.timeStamp - _this.__t.start > 30 && e.timeStamp - _this.__t.start < 1000 && this.clickabled) {
//                     setTimeout(function() {
//                         if (!!fn) {
//                             fn(e);
//                         }
//                     }, 0);
//                 }
//             },
//             click: function(e) {
//                 e.stopPropagation();
//                 fn && fn(e);
//             }
//         };
//         if (!!navigator.userAgent.match(/AppleWebKit.*Mobile.*/)) {
//             this.addEventListener('touchstart', this.tevent.tstart, false);
//             this.addEventListener('touchmove', this.tevent.tmove, false);
//             this.addEventListener('touchend', this.tevent.tend, false);
//         } else {
//             this.addEventListener('click', this.tevent.click, false);
//         }
//     };
//     HTMLElement.prototype.untap = function() {
//         if (!!navigator.userAgent.match(/AppleWebKit.*Mobile.*/) && !!this.tevent ) {
//             this.tevent.tstart && this.removeEventListener('touchstart', this.tevent.tstart, false);
//             this.tevent.tmove && this.removeEventListener('touchmove', this.tevent.tmove, false);
//             this.tevent.tend && this.removeEventListener('touchend', this.tevent.tend, false);
//         } else if(!!this.tevent) {
//             this.tevent.click && this.removeEventListener('click', this.tevent.click, false);
//         }
//     }
// }


index.js
'use strict';
var Eventor = require('eventor');
var PickerDialog = require('./src/picker-dialog.js');
var PickerCore = require('picker-core');
var F_tap = require('./src/simulate-click.js');
require('./src/picker.css');

function getNode(tag, classname, value) {
    var node = document.createElement(tag);
    node.className = classname || "";
    node.innerHTML = value || "";
    return node;
}
var Picker = function(option) {
    var _this = this;
    if( Object.prototype.toString.call(option) === '[object Object]' ){
        this.params = {
            "itemsNumber": option.itemsNumber || 7,
            "itemHeight": option.itemHeight || 30,
            "cols": option.cols || [],
            "input": option.input || "",
            "inputContainer": document.querySelector(option.input) || "",
            "toolbarTemplate": option.toolbarTemplate || ('<div class="picker-toolbar">' +
                '<a href="javascript:void(0);" class="picker-toolbar-right">Done</a>' +
                '</div>'),
            "formatValue": option.formatValue || function(picker, values) {
                return values.join(" ");
            },
            "after": option.after,
            "_open": function(){ // 内部打开事件
                _this.inputSelf.value = _this.params.formatValue(_this, _this._getValues());
            },
            "_close": function(){

            }
        };
        this.input = document.querySelector(option.input);
        this.inputSelf = this.input;
        if (!!option.container) {
            this.params.container = option.container || "";
            this.container = document.querySelector(option.container);
        }
        if (!!option.onOpen) this.params.onOpen = option.onOpen;
        if (!!option.onClose) this.params.onClose = option.onClose;
        if (!!option.onChange) this.params.onChange = option.onChange;

        option = null;
    };
    if (this.isError()) return false;
    this.init();
    return this;
};
Eventor.mixTo(Picker);
Picker.prototype.isError = function() {
    if (!this.input) {
        console.error('input对应的dom对象不存在');
        return true;
    }
    if( !!this.params.container && !this.container ){
        console.error('container对应的dom对象不存在');
        return true;
    }
    return false;
};
Picker.prototype.init = function() {
    this.input.setAttribute('readonly', 'readonly');
    this.input.style.webkitTapHighlightColor = 'rgba(0,0,0,0)';
    // this._hackInputFocus();
    if (this.container) {
        this.dialog = {};
        this.dialog.container = this.container;
        this.dialog.container.innerHTML = '<div class="picker-cols"></div>';
        this._wrap = this.dialog.container.querySelector('.picker-cols');
        this.initPickerCore();
    } else {
        this.initDialog();
        this.initPicker();
        this.initPickerCore();
        this._bindEventToolbar();
    }
    this.params.after && this.params.after(this);
    this._getValues();
    return this;
};
Picker.prototype.destroy = function() {
    if ((this.container.classList + "").indexOf("modal-in") > 0) this.dialog.hide();
    if (!!this.params.container) this.container.innerHTML = "";
    else this.container.remove();
    // this.input.untap();
    F_tap.untap( this.input );
    clearTimeout(this.timeout);
    for (var i in this) {
        delete this[i];
    }
};
Picker.prototype.initPicker = function() {
    this._wrap = this.container.querySelector('.picker-cols');
    return this;
};
Picker.prototype.initDialog = function() {
    this.dialog = new PickerDialog(this.params);
    //this._wrap = this.dialog.container;
    if (!this.params.container) {
        this.container = this.dialog.container;
        this.container.innerHTML = this.params.toolbarTemplate + '<div class="picker-cols"></div>';
    }
    return this;
};
Picker.prototype.close = function() {
    return this.dialog.hide(), this;
};
Picker.prototype.open = function() {
    // this.inputSelf.value = this.params.formatValue(this, this._getValues());
    return this.dialog.show(), this;
};
Picker.prototype.initPickerCore = function() {
    var _this = this;
    this.cols = [];
    var _cols = this.params.cols;
    for (var i = 0; i < _cols.length; i++) {
        if (_cols[i].divider === true) {
            this._wrap.appendChild(getNode("div", "basescroller-divider", _cols[i].content));

        } else {
            if (!_cols[i].values) continue;
            //if( !!_cols[i].values ){
            if (!!_cols[i].displayValues) {
                var _data = [];
                for (var j = 0; j < _cols[i].values.length; j++) {
                    _data.push({
                        "name": _cols[i].displayValues[j] || _cols[i].values[j],
                        "value": _cols[i].values[j]
                    });
                }
            } else {
                var _data = _cols[i].values;
            }
            var option = {
                container: this._wrap,
                data: _data,
                itemHeight: this.params.itemHeight,
                itemsNumber: this.params.itemsNumber,
                selectedCallback: (function() {
                    var _i = i;
                    return function(_new, _old) {
                        if (!!_cols[_i].onChange) _cols[_i].onChange(_this, _new, _old);
                        else if (!!_this.params.onChange) _this.params.onChange(_this, _new, _old);
                        _this.inputSelf.value = _this.params.formatValue(this, _this._getValues());
                    }
                }())
            };
            var _temp = new PickerCore(option);
            _temp.container = _temp.scrollerComponent;
            _temp.items = _temp.container.querySelectorAll(".basescroller-item");
            delete _temp.scrollerComponent;
            _temp.values = _cols[i].values;
            _temp.replaceValues = function(values, displayValues) {
                _this._InitCssPickerCore();
                if (!displayValues) this.render(values);
                else {
                    var arr = [];
                    for (var i = 0; i < values.length; i++) {
                        arr.push({
                            "name": displayValues[i] || values[i],
                            "value": values[i]
                        });
                    }
                    this.render(arr);
                }
            };
            this.cols.push(_temp);

            //}
        }
    }
    // picker-core 样式


    // picker-core 样式
    this._wrap.appendChild(getNode("div", "basescroller-current-indicator-all"));
    this._component = this.container.querySelectorAll('.basescroller-component');
    // this.picker.scroller = this.dialog.container.querySelectorAll('.basescroller-scroller');
    this._InitCssPickerCore();
    return this;
};
Picker.prototype._InitCssPickerCore = function() {
    var _this = this;
    // 设置选择项的width、text-align
    _this._setCssWidth();
    this.timeout = setTimeout(function() {
        if( !!_this.params ){
            _this._setCssWidth();
            _this._wrap.style.height = _this.params.itemsNumber * _this.params.itemHeight + "px";

            var _item = _this._wrap.querySelectorAll(".basescroller-item");
            var _indicator = _this._wrap.querySelectorAll(".basescroller-current-indicator");
            var _indicatorAll = _this._wrap.querySelector(".basescroller-current-indicator-all");
            for (var i = 0; i < _this._component.length; i++) {
                _this._component[i].style.height = _this.params.itemsNumber * _this.params.itemHeight + "px";
            }
            for (var i = 0; i < _indicator.length; i++) {
                _indicator[i].style.height = _this.params.itemHeight - 2 + "px";
                _indicator[i].style.top = _this.params.itemHeight * (_this.params.itemsNumber - 1) / 2 + "px";
            }
            _indicatorAll.style.top = _this.params.itemHeight * (_this.params.itemsNumber - 1) / 2 + "px";
            _indicatorAll.style.height = _this.params.itemHeight - 2 + "px";
            for (var i = 0; i < _item.length; i++) {
                _item[i].style.height = _this.params.itemHeight + "px";
                _item[i].style.lineHeight = _this.params.itemHeight + "px";
            }
        }
    }, 100);
    return this;
};
Picker.prototype._setCssWidth = function() {
    for (var i = 0; i < this._component.length; i++) {
        var componentWidth = this._component[i].querySelector(".basescroller-scroller");
        componentWidth = !!componentWidth ? componentWidth.offsetWidth : this._component[i].offsetWidth;
        this._component[i].style.width = !!this.params.cols[i].width ? this.params.cols[i].width + 'px' : componentWidth + 'px';
        this._component[i].style.textAlign = this.params.cols[i].textAlign || 'center';
    }
}
Picker.prototype.setValue = function(values) {
    // 设置新值给当前的选项列
    if (Object.prototype.toString.call(values) == "[object Array]") {
        for (var i = 0; i < this.cols.length; i++) {
            if (!!values[i]) {
                this.cols[i].select(values[i]);
            }
        }
    } else console.error("setValue的参数必须是数组");
}
Picker.prototype._bindEventToolbar = function() {
    // 设置左右toolbar按钮的事件
    var _this = this;
    var toolbarRight = _this.container.querySelector('.picker-toolbar-right');
    if (toolbarRight) {
        _this.on('toolbarRight', function(e) {
            _this.close();
        });
        toolbarRight.click = function(e) {
            _this.emit('toolbarRight', e);
        };
        F_tap.tap( toolbarRight, toolbarRight.click );
        // toolbarRight.tap(toolbarRight.click);
    };
};
Picker.prototype._bindEventInput11111 = function() {
    // 设置input被触发的事件
    var _this = this;
    F_tap.untap( this.input );
    F_tap.tap( this.input, function(){
        _this.open();
        // 如果input被picker遮挡到，则滚动input至可视区域
        (function() {
            var clientHeight = document.documentElement.clientHeight; //  浏览器高度
            var scrollTop = document.body.scrollTop; // 滚动高度
            var pickerHeight = _this.dialog.container.offsetHeight; // picker高度
            var inputTop = _this.input.offsetTop; // input相对body高度
            var inputHeight = _this.input.offsetHeight; // input高度
            if (inputTop - scrollTop + inputHeight > clientHeight - pickerHeight || inputTop - scrollTop + inputHeight < inputHeight) {
                _this.scrollAnimate(inputTop - (clientHeight - pickerHeight) / 2, 400);
            }
        }());
    } );
    // this.input.untap();
    // this.input.tap(function() {
    //     _this.open();
    //     // 如果input被picker遮挡到，则滚动input至可视区域
    //     (function() {
    //         var clientHeight = document.documentElement.clientHeight; //  浏览器高度
    //         var scrollTop = document.body.scrollTop; // 滚动高度
    //         var pickerHeight = _this.dialog.container.offsetHeight; // picker高度
    //         var inputTop = _this.input.offsetTop; // input相对body高度
    //         var inputHeight = _this.input.offsetHeight; // input高度
    //         if (inputTop - scrollTop + inputHeight > clientHeight - pickerHeight || inputTop - scrollTop + inputHeight < inputHeight) {
    //             _this.scrollAnimate(inputTop - (clientHeight - pickerHeight) / 2, 400);
    //         }
    //     }());
    // });
    F_tap.tap( this.params.inputContainer, function(){
        _this.open();
        // 如果input被picker遮挡到，则滚动input至可视区域
        (function() {
            var clientHeight = document.documentElement.clientHeight; //  浏览器高度
            var scrollTop = document.body.scrollTop; // 滚动高度
            var pickerHeight = _this.dialog.container.offsetHeight; // picker高度
            var inputTop = _this.input.offsetTop; // input相对body高度
            var inputHeight = _this.input.offsetHeight; // input高度
            if (inputTop - scrollTop + inputHeight > clientHeight - pickerHeight || inputTop - scrollTop + inputHeight < inputHeight) {
                _this.scrollAnimate(inputTop - (clientHeight - pickerHeight) / 2, 400);
            }
        }());
    } );
    // this.params.inputContainer.tap(function() {
    //     _this.open();
    //     // 如果input被picker遮挡到，则滚动input至可视区域
    //     (function() {
    //         var clientHeight = document.documentElement.clientHeight; //  浏览器高度
    //         var scrollTop = document.body.scrollTop; // 滚动高度
    //         var pickerHeight = _this.dialog.container.offsetHeight; // picker高度
    //         var inputTop = _this.input.offsetTop; // input相对body高度
    //         var inputHeight = _this.input.offsetHeight; // input高度
    //         if (inputTop - scrollTop + inputHeight > clientHeight - pickerHeight || inputTop - scrollTop + inputHeight < inputHeight) {
    //             _this.scrollAnimate(inputTop - (clientHeight - pickerHeight) / 2, 400);
    //         }
    //     }());
    // });
    return this;
};
Picker.prototype._getValues = function() {
    var _temp = [];
    for (var i = 0; i < this.cols.length; i++) {
        if (this.cols[i].getValue) _temp.push(this.cols[i].getValue());
    };
    this.values = _temp;
    return _temp;
}
module.exports = Picker;
